### ISSUE TITLE
Schema::Rule should inherit from BasicObject

### ISSUE THREAD

**Issue by solnic** [dryrb/dry-validation#50]
Otherwise high-level rules won't work with predicates that are also `Object' methods, ie `eql?`

**Comment by fran-worley** [comment]
Might be unrelated, but I can't get this to work:

```ruby
    key(:login) { |login| login.bool? }
    key(:email) { |email| email.none? | email.format?(/\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\Z/i) }
    
    rule(:require_email) do
      rule(:login).true?.then(rule(:email).filled?)
    end

    def true?(value)
      raise value.inspect
      value === true
    end
```
The key definitions work fine, but the high level rule never seems to get executed on the basis that my exception is never raised.

**Comment by solnic** [comment]
Could you show me a full example?

**Comment by fran-worley** [comment]
In theory, this: 

```ruby
class UserSchema < Dry::Validation::Schema
  key(:login) { |login| login.bool? }
  key(:email) { |email| email.none? | email.format?(/\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\Z/i) }

  rule(:require_email) do
      rule(:login).true?.then(rule(:email).filled?)
  end

  def true?(value)
    value === true
  end
end
schema = UserSchema.new
errors = schema.call(email: nil, login: true)
```

Should raise an error on email as it should be filled when login === true
